### DISCUSSION TITLE
ImageResizing to MultipleDimension Using ImageMagick

### DISCUSSION THREAD

**Discussion by programmerarunkumar** [ImageMagick/ImageMagick#5212] [Help]
Hi,

Im using MagickWand Method to resize the image

Im using the below code to resize the single image to multiple dimensions

```
MagickWand *magick_wand=NewMagickWand();
MagickReadImage(magick_wand,originalImage);
int count = 0;
while(count<imageCount){
        MagickResizeImage(magick_wand, widthList[count], heightList[count], LanczosFilter);
        MagickWriteImages(magick_wand, resizeImage[count], MagickTrue);
        count++;
}
magick_wand=DestroyMagickWand(magick_wand);
````

Is it okay to do it like this?
correct me if the above code will lead to any issues 

As I have tested this code, I haven't faced any issue, However I needed the clarification.

**Comment by snibgo** [comment]
> correct me if the above code will lead to any issues

You are resizing an image, then resizing the result, then resizing the result of that.

In general, this will reduce image quality at each resize. The results will be better if each resize is of the original image, not of the previous resize.

**Comment by programmerarunkumar** [comment]
Im trying to resize an image to different Dimensions

Im reading the original image using MagickReadImage before the While Loop
Inside the while loop im resizing the image to respective dimension and writing in respective file


Then how come this can happen 
> You are resizing an image, then resizing the result, then resizing the result of that.

can you explain little bit deeper

**Comment by snibgo** [comment]
For example, suppose your input image is 200x200 pixels, and you have two resizes. Suppose the first resize is to 1x1, and the second is to 100x100.

With your code, the first resize will convert from 200x200 to 1x1, ie just one pixel. The second resize will convert from 1x1 to 100x100. The second output will be a single colour, and won't look like the input image.

Instead, I suggest the second resize should convert from the input image, so it will convert from 200x200 to 100x100. The result will look like the input image, but smaller.

**Reply by programmerarunkumar** [reply]
ImageMagick will change the original image in the magickwand after the resize. Hence the second resize will read the previous result ?

**Reply by programmerarunkumar** [reply]
can someone clarify this ?

**Reply by fmw42** [reply]
Clarify what?  Please explain your confusion.  Best to resize from the original each time, or if downsizing one resolution to the next, do it from highest resolution to lowest resolution. Do not downsize and then upsize.

**Reply by programmerarunkumar** [reply]
My doubt is I haven't changed the original image during resize, then how it is using the previous resized image for next resize ?

**Reply by fmw42** [reply]
If you use +clone for each step, it will resize from the previous image. If you use the original image each time or use -clone 0, then it will resize from the original image.

**Comment by snibgo** [comment]
> My doubt is I haven't changed the original image during resize

The program fragment you show does exactly that. "MagickResizeImage()" changes the size of the image.

**Reply by programmerarunkumar** [reply]
MagickResizeImage() will change my original image ?

**Comment by snibgo** [comment]
What "original image"? In your code, originalImage is (I suppose) a string containing a filename. That file contains, I suppose, a single image.

Your code reads an image into magick_wand. Then it resizes the image in magick_wand, and writes the image in magick_wand to a file.

Then, depending on imageCount, it may again resize the image in magick_wand, and again write the image in magick_wand to a file.

So the resizing is cumulative. The image is resized, and the result is resized, and *that* result is resized, and so on.

I suggest you try the experiment, perhaps with a simplified version of your program. Choose any image that has some detail. Resize that image to 1x1 pixel, then resize the result to the same size as the original. The final result will be different to the original image.

**Reply by programmerarunkumar** [reply]
> What "original image"? 

> In your code, originalImage is (I suppose) a string containing a filename. That file contains, I suppose, a single  image.

Yes correct.

My doubts:
When resized image will replace the original image in the magick wand ?
How original image and resized image stored/handled in magickwand ?

**Comment by snibgo** [comment]
> When resized image will replace the original image in the magick wand ?
How original image and resized image stored/handled in magickwand ?

A MagickWand doesn't have an "original image". It may contain many images. In your case, I suppose it has only one image. If you resize that image, it is immediately resized.

Resizing an image multiple times will lose quality. This is like making a photocopy of a photocopy of a photocopy.

It is better to read an image and make a copy, resize the copy, write it and delete it. Then make another copy of the read image, resize and write and delete the copy.

There are various methods for making a copy. For example, we can copy the entire MagickWand:
```
  clone_wand = CloneMagickWand (magick_wand);
```

**Reply by programmerarunkumar** [reply]
> clone_wand = CloneMagickWand (magick_wand);

MagickWand *clone_wand;
MagickWand *magick_wand=NewMagickWand();
MagickReadImage(magick_wand,originalImage);
int count = 0;
while(count<imageCount){
        clone_wand = CloneMagickWand(magick_wand);
        MagickResizeImage(clone_wand, widthList[count], heightList[count], LanczosFilter);
        MagickWriteImages(clone_wand, resizeImage[count], MagickTrue);
        clone_wand=DestroyMagickWand(clone_wand);
        count++;
}
magick_wand=DestroyMagickWand(magick_wand);

Can I use like this ?

**Comment by snibgo** [comment]
Yes.
